src/StringTools/StringPool.cs

```diff
@@ -0,0 +1,294 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+
+namespace Microsoft.NET.StringTools;
+
+/// <summary>
+/// Facilitates pooling of strings constructed from <c>ReadOnlySpan&lt;char&gt;</c> values,
+/// providing zero-allocation lookup for strings that have been seen before.
+/// </summary>
+/// <remarks>
+/// Uses ordinal string comparison.
+/// </remarks>
+public sealed class StringPool
+{
+    private int[]? _buckets;
+    private Slot[]? _slots;
+    private int _count;
```

Comment:
We have a similar string pool data structure implemented in StringTools called `WeakStringCache`. The main difference is that it's "weak", so does not guarantee that strings can be recovered if nothing else keeps them alive. It would be interesting to measure how well it would work in the solution file parsing scenario - could be done simply by making the `Intern` method here call `Strings.WeakString`.

I'm also curious about the implementation choices. To avoid reimplementing a hashtable, in `WeakStringCache` we chose to internally keep strings keyed by their hashcode in (essentially) `Dictionary<int, string>`. I can see that a proper implementation as done here may be more efficient but I suspect that it's not by much and the cost of maintaining a hashtable here is somewhat unfortunate. Do you think it would be possible to benchmark the simpler approach of calculating a hash and than relying on `Dictionary<int, string>`? Or `Dictionary<int, object>` with a polymorphic value if we want to support collisions.